metrics/performance/first-meaningful-paint.js

```diff
@@ -18,49 +18,61 @@
 
 const DevtoolsTimelineModel = require('devtools-timeline-model');
 
+const FAILURE_ERROR = new Error('Navigation and first paint timings not found');
+
 class FirstMeaningfulPaint {
+
   /**
    * @param {!Array<!Object>} traceData
    */
   static parse(traceData) {
     return new Promise((resolve, reject) => {
+      if (!traceData || !Array.isArray(traceData)) {
+        return reject(FAILURE_ERROR);
+      }
+
       const model = new DevtoolsTimelineModel(traceData);
       const events = model.timelineModel().mainThreadEvents();
-      const ret = {};
 
-      const navStartEvent = events
-          .filter(FirstMeaningfulPaint._filterEventsForNavStart);
-      const firstTextPaintEvent = events
-          .filter(FirstMeaningfulPaint._filterEventsForFirstTextPaint);
+      // Identify the frameID of the main frame
+      const startedInPage = model.tracingModel().devToolsMetadataEvents()
+        .filter(e => e.name === 'TracingStartedInPage')
+        .sort((a, b) => a.startTime - b.startTime)
+        .slice(-1);
+      const frameID = startedInPage[0].args.data.page;
 
-      if (firstTextPaintEvent.length && navStartEvent.length) {
-        ret.duration = firstTextPaintEvent[0].startTime - navStartEvent[0].startTime;
-      } else {
-        ret.err = new Error('First meaningful paint metric not found');
-      }
+      // Find the start of navigation and our meaningful paint
+      const userTiming = events
+        .filter(e => e.categoriesString.includes('blink.user_timing'))
+        // Events can be unsorted, so we put in ascending order.
+        .sort((a, b) => a.startTime - b.startTime);
 
-      resolve(ret);
-    });
-  }
+      // navigationStart == the network begins fetching the page URL
+      // CommitLoad == the first bytes of HTML are returned and Chrome considers
+      //   the navigation a success. A 'isMainFrame' boolean is attached to those events
+      //   However, that flag may be incorrect now, so we're ignoring it.
+      const navStart = userTiming.filter(e => {
+        return e.name === 'navigationStart' && e.args.frame === frameID;
+      }).slice(-1);
```

Comment:
again you can reverse the order if you want to drop the `slice`s and do
```
const navStart = userTiming.filter(e => {
  return e.name === 'navigationStart' && e.args.frame === frameID;
})[0];
```
but knowing you, you may prefer the `slice` :)